lib/elixir/lib/task/supervised.ex

```diff
@@ -210,27 +215,39 @@ defmodule Task.Supervised do
         waiting =
           case waiting do
             %{^position => {_, {:ok, _} = ok}} -> Map.put(waiting, position, {nil, ok})
-            %{^position => {_, :running}} -> Map.put(waiting, position, {nil, {:exit, reason}})
+            %{^position => {_, _timer_ref, :running}} -> Map.put(waiting, position, {nil, {:exit, reason}})
+            %{^position => {_, :timed_out}} -> Map.put(waiting, position, {nil, {:exit, :killed}})
           end
         stream_deliver({:cont, acc}, max + 1, spawned, delivered, waiting, next,
-                       reducer, monitor_pid, monitor_ref, timeout)
+                       reducer, monitor_pid, monitor_ref, timeout, on_timeout)
+
+      # The task at position "position" didn't reply in time, we'll decide what
+      # to do based on the :on_timeout option. It may also be that the task
+      # replied (we have {:ok, _} in "waiting") but it didn't die yet (we didn't
+      # get :down); in that case, we do nothing.
+      {:timed_out, {^monitor_ref, position}} ->
+        case on_timeout do
+          :exit ->
+            stream_close(monitor_pid, monitor_ref, timeout)
```

Comment:
There can be pending timers in our `waiting` map, we would need to cancel all these timers on `stream_close`, as the Caller might catch the `exit/1`.